Nitrox.Launcher/ViewModels/ServersViewModel.cs

```diff
@@ -0,0 +1,230 @@
+    }
+
+    private async Task GetSavesOnDiskAsync()
+    {
+        try
+        {
+            Directory.CreateDirectory(keyValueStore.GetSavesFolderDir());
+
+            Dictionary<string, (ServerEntry Data, bool HasFiles)> serversOnDisk;
+            lock (serversLock)
+            {
+                serversOnDisk = Servers.ToDictionary(entry => entry.Name, entry => (entry, false));
+            }
+            foreach (string saveDir in Directory.EnumerateDirectories(keyValueStore.GetSavesFolderDir()))
+            {
+                try
+                {
+                    if (serversOnDisk.TryGetValue(Path.GetFileName(saveDir), out (ServerEntry Data, bool _) server))
+                    {
+                        // This server has files, so don't filter it away from server list.
+                        serversOnDisk[Path.GetFileName(saveDir)] = (server.Data, true);
+                        continue;
+                    }
+                    ServerEntry entryFromDir = await Task.Run(() => ServerEntry.FromDirectory(saveDir));
+                    if (entryFromDir != null)
+                    {
+                        serversOnDisk.Add(entryFromDir.Name, (entryFromDir, true));
+                    }
+                    loggedErrorDirectories.Remove(saveDir);
+                }
+                catch (Exception ex)
+                {
+                    if (loggedErrorDirectories.Add(saveDir)) // Only log once per directory to prevent log spam
+                    {
+                        Log.Error(ex, $"Error while initializing save from directory \"{saveDir}\". Skipping...");
+                    }
+                }
+            }
+
+            lock (serversLock)
+            {
+                Servers = [..serversOnDisk.Values.Where(server => server.HasFiles).Select(server => server.Data).OrderByDescending(entry => entry.LastAccessedTime)];
+            }
+        }
+        catch (Exception ex)
+        {
+            Log.Error(ex, "Error while getting saves");
+            await dialogService.ShowErrorAsync(ex, "Error while getting saves");
+        }
+    }
+
+    private async Task WatchServersAsync(CancellationToken cancellationToken = default)
+    {
+        watcher = new FileSystemWatcher
+        {
+            Path = keyValueStore.GetSavesFolderDir(),
+            NotifyFilter = NotifyFilters.DirectoryName | NotifyFilters.LastWrite | NotifyFilters.Size,
+            Filter = "*.*",
+            IncludeSubdirectories = true
+        };
+        watcher.Changed += OnDirectoryChanged;
+        watcher.Created += OnDirectoryChanged;
+        watcher.Deleted += OnDirectoryChanged;
+        watcher.Renamed += OnDirectoryChanged;
+
+        await Task.Run(async () =>
+        {
+            watcher.EnableRaisingEvents = true; // Slowish (~2ms) - Moved into Task.Run.
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
I would rather check for IsCancellationRequested to gracefully shut down the Task, we don't really need to take advantage of the exception system here

Reply:
This is a .NET API thing and how `Task.Delay` works too. If `IsCancellationRequested` is used, the task will not have a canceled state as it exited normally, so it's better to throw. We could use `Task.FromCanceled` but then you'd still need to `try...catch` anyway in case a `Task.Delay` was canceled.